script/tool/lib/src/pubspec_check_command.dart

```diff
@@ -542,17 +542,25 @@ class PubspecCheckCommand extends PackageLoopingCommand {
   // there are any that aren't allowed.
   String? _checkDependencies(Pubspec pubspec) {
     final Set<String> badDependencies = <String>{};
+    // Shipped dependencies.
     for (final Map<String, Dependency> dependencies
-        in <Map<String, Dependency>>[
-      pubspec.dependencies,
-      pubspec.devDependencies
-    ]) {
+        in <Map<String, Dependency>>[pubspec.dependencies]) {
       dependencies.forEach((String name, Dependency dependency) {
         if (!_shouldAllowDependency(name, dependency)) {
           badDependencies.add(name);
         }
       });
     }
+    // Dev dependencies
+    for (final Map<String, Dependency> dependencies
+        in <Map<String, Dependency>>[pubspec.devDependencies]) {
+      dependencies.forEach((String name, Dependency dependency) {
+        if (!_shouldAllowDevDependency(name, dependency)) {
+          badDependencies.add(name);
+        }
+      });
+    }
+
```

Comment:
I think it would be less code/complexity if you left the old structure in place and then just added a second section here:
```dart
// Ensure that dev-only dependencies aren't in `dependencies`.
const List<String> devOnlyDependencies = <String>['integration_test'];
pubspec.dependencies.forEach((String name, Dependency dependency) {
  if (devOnlyDependencies.contains(name)) {
    badDependencies.add(name);
  }
});
```

(In fact, that last loop probably condenses to something like `badDependencies.addAll(pubspec.dependencies.keys.where(devOnlyDependencies.contains));`

Reply:
Done, Left in the loop form though because I find it easier to read.